src/App/Fossa/Analyze.hs

```diff
@@ -158,16 +158,24 @@ tryUploadContributors baseDir baseUrl apiKey locator = do
   contributors <- fetchGitContributors baseDir
   uploadContributors baseUrl apiKey locator contributors
 
-fossaProjectUrl :: URI -> Text -> Text -> Text
-fossaProjectUrl baseUrl rawLocator branch = URI.render baseUrl <> "projects/" <> encodedProject <> "/refs/branch/" <> branch <> "/" <> encodedRevision
+-- This url can have a two forms (Core may allow more, but we don't care here):
+--    https://<fossa host>/projects/<project>/
+--    https://<fossa host>/projects/<project>/refs/branch/<branch>/<revision>
+fossaProjectUrl :: URI -> Text -> ProjectRevision -> Text
+fossaProjectUrl baseUrl rawLocator revision = URI.render baseUrl <> "projects/" <> encodedProject <> buildSelector
   where
     Locator{locatorFetcher, locatorProject, locatorRevision} = parseLocator rawLocator
 
     underBS :: (ByteString -> ByteString) -> Text -> Text
     underBS f = TE.decodeUtf8 . f . TE.encodeUtf8
-
-    encodedProject = underBS (urlEncode True) (locatorFetcher <> "+" <> locatorProject)
-    encodedRevision = underBS (urlEncode True) (fromMaybe "" locatorRevision)
+    urlEncode' = underBS (urlEncode True)
+
+    encodedProject = urlEncode' (locatorFetcher <> "+" <> locatorProject)
+    encodedRevision = urlEncode' (fromMaybe "" locatorRevision)
+    -- | buildSelector is empty string unless we have a real branch to work with.
+    buildSelector = fromMaybe "" $ do
+      branch <- projectBranch revision
+      Just $ "/refs/branch/" <> urlEncode' branch <> "/" <> encodedRevision
 
```

Comment:
I'm a little iffy on this `do` notation here.  Is there a better way to write this, or are the other ways more confusing?

Reply:
I think this way is perfect. It's immediately clear to me what this is doing. the other common way is to use `maybe :: b -> (a -> b) -> Maybe a -> b`:
```hs
buildSelector = maybe ""
  (\branch -> "/refs/branch/" <> urlEncode' branch <> "/" <> encodedRevision)
  $ projectBranch revision
```